src/core/stdcpp/typeinfo.d

```diff
@@ -0,0 +1,146 @@
+          private:
+            //this(const type_info rhs);
+            //type_info operator=(const type_info rhs);
+        }
+
+        class bad_cast : core.stdcpp.exception.std.exception
+        {
+            this() nothrow { }
+            this(const bad_cast) nothrow { }
+            //bad_cast operator=(const bad_cast) nothrow { return this; }
+            //virtual ~this() nothrow;
+            override const(char)* what() const nothrow;
+        }
+
+        class bad_typeid : core.stdcpp.exception.std.exception
+        {
+            this() nothrow { }
+            this(const bad_typeid) nothrow { }
+            //bad_typeid operator=(const bad_typeid) nothrow { return this; }
+            //virtual ~this() nothrow;
+            override const (char)* what() const nothrow;
+        }
+    }
+}
+else version (CRuntime_Microsoft)
+{
+    import core.stdcpp.exception;
+
+    struct __type_info_node
+    {
+        void* _MemPtr;
+        __type_info_node* _Next;
+    }
+
+    extern __gshared __type_info_node __type_info_root_node;
+
+    extern (C++, std)
+    {
+        class type_info
+        {
+
+          public:
+            //virtual ~this();
+            void dtor() { }     // reserve slot in vtbl[]
+            //bool operator==(const type_info rhs) const;
+            //bool operator!=(const type_info rhs) const;
+            final bool before(const type_info rhs) const;
+            final const(char)* name(__type_info_node* p = &__type_info_root_node) const;
+
+          private:
+            void* pdata;
+            char _name[1];
+            //this(const type_info rhs);
+            //type_info operator=(const type_info rhs);
+        }
+
+        class bad_cast : core.stdcpp.exception.std.exception
+        {
+            this(const(char)* msg = "bad cast") { }
+            this(const bad_cast) { }
+            //virtual ~this();
+        }
+
+        class bad_typeid : core.stdcpp.exception.std.exception
+        {
+            this(const(char)* msg = "bad typeid") { }
+            this(const bad_typeid) { }
+            //virtual ~this();
+        }
+    }
+}
+else version (CRuntime_Glibc)
+{
+    import core.stdcpp.exception;
+
+    extern (C++, __cxxabiv1)
+    {
+        class __class_type_info;
+    }
+
+    extern (C++, std)
+    {
+        class type_info
+        {
+            void dtor1();                           // consume destructor slot in vtbl[]
+            void dtor2();                           // consume destructor slot in vtbl[]
+            final const(char)* name();
+            final bool before(const type_info) const;
+            //bool operator==(const type_info) const;
+            bool __is_pointer_p() const;
+            bool __is_function_p() const;
+            bool __do_catch(const CppTypeInfo, void**, uint) const;
+            bool __do_upcast(const __cxxabiv1.__class_type_info, void**) const;
+
+            const(char)* _name;
+            this(const(char)*);
```

Comment:
If C++ code uses RTTI, and we pass our D-constructed object, it may cause strange unexpected error:
```
//C++
void thridLibraryFunction(std::exception *ex)
{
    if (std::bad_typeid *ebti = dynamic_cast<std::bad_typeid>(ex))
    {
        //do something
    }
    else
    {
         //do something else
    }
}


//D code:

extern(C++) void thridLibraryFunction(std.exception ex);

extern(C++) class d_exception : std.exception
{
     override const (char)* what() const nothrow
     {
          return "From D with love";
     }
}

void main()
{
    thridLibraryFunction(new d_exception()); //unexpected segfault on dynamic_cast
}
```

>We could implement C++ RTTI

It may be difficult for VC++, because VC++ rtti is undocumented.
clang++ probably resolves this trouble and I have reverse-enginered scheme of VC++ RTTI, but do we want to take on this headache?